### PR TITLE
add a metadata refresh command to tuftool

### PR DESCRIPTION
**Pull request by webern**
*Issue #, if available:*

N/A

*Description of changes:*

* Adds a command to the tuftool which pulls metadata files, updates their expiration dates and version numbers, and re-signs them.
* Adds code paths that check and error if no signing keys are available.
* Fixes an issue where the `create_verify_repo` test was not asserting what we wanted it. Renamed this test to `create_command`.
* Includes a `refresh_command` test which tests the `refresh` command.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tuftool/src/refresh.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                    },
+                },
+                _extra: HashMap::new(),
+            },
+            self.snapshot_version,
+            "snapshot.json",
+            &rng,
+        )?;
+
+        // create, sign, and write out the timestamp file
```
Comment (webern) [inline tuftool/src/refresh.rs]: comma use is inconsistent with line 132. choose one.

File: tuftool/src/refresh.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        metadata::write_metadata(
```
Comment (webern) [inline tuftool/src/refresh.rs]: rename function to `write` to avoid duplicating the word metadata?

File: .gitignore [hunk 3]
```diff
@@ -1,2 +1,3 @@
 /target
 **/*.rs.bk
+.idea
```
Comment (iliana) [inline .gitignore]: I've been avoiding adding editor-specific files to .gitignore. [Consider setting up a global .gitignore file.](https://help.github.com/en/github/using-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer)

File: tuftool/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.3.0] - 2020-02-05
```
Comment (iliana) [inline tuftool/CHANGELOG.md]:
```suggestion
## [Unreleased]
```

See the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

File: tuftool/CHANGELOG.md [hunk 5]
```diff
@@ -30,6 +34,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Everything!
 
+[0.3.0]: https://github.com/awslabs/tough/compare/tuftool-v0.2.0...tuftool-v0.3.0
```
Comment (iliana) [inline tuftool/CHANGELOG.md]:
```suggestion
[Unreleased]: https://github.com/awslabs/tough/compare/tuftool-v0.2.0...HEAD
```

File: tuftool/src/error.rs [hunk 6]
```diff
@@ -151,6 +151,9 @@ pub(crate) enum Error {
         backtrace: Backtrace,
     },
 
+    #[snafu(display("Unable to match any of the provided keys with root.json"))]
+    KeysNotFoundInRoot {},
```
Comment (iliana) [inline tuftool/src/error.rs]: Please add a `Backtrace` here :)

File: tuftool/src/create.rs [hunk 7]
```diff
@@ -119,7 +116,7 @@ struct CreateProcess<'a> {
 }
 
 impl<'a> CreateProcess<'a> {
-    fn run(self) -> Result<()> {
+    fn run(self) -> crate::error::Result<()> {
```
Comment (iliana) [inline tuftool/src/create.rs]: This is still imported.

File: tuftool/src/digest.rs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+///
+/// # Arguments
+///
+/// * `path` - The filepath to a root.json file.
+///
+///  # Return
+///
+/// * Either an Error, or a tuple consisting of `Root`, `sha256` digest, and `length`.
+///
+pub(crate) fn digest_root_file(path: &PathBuf) -> Result<(Root, [u8; SHA256_OUTPUT_LEN], u64)> {
```
Comment (iliana) [inline tuftool/src/digest.rs]: Instead of returning tuples from methods, a struct is almost always better.

File: tuftool/src/metadata.rs [hunk 9]
```diff
@@ -0,0 +1,81 @@
+use crate::error::{self, Result};
```
Comment (iliana) [inline tuftool/src/metadata.rs:4]: New files need to have a copyright header.

```rust
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT OR Apache-2.0
```

**Review by iliana — COMMENTED** [review]
Thank you!

This is a first review pass. I'll need to do another pass to understand some of the refactoring done here.

**Comment by iliana** [comment]
You can add `#![allow(clippy::must_use_candidate)]` to main.rs, I don't think I like that particular lint.

File: .gitignore (hunk 3 above)
Comment (webern) [reply .gitignore]: I check in non-volatile .idea files in a lot of my other projects so a global would be problematic. But I can manually remove it from PRs when I open them.

File: tuftool/src/refresh.rs [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    fn refresh<'a, T: Transport>(&self, repository: &Repository<'a, T>) -> Result<()> {
+        // clone the targets.json file but with a new version number and expiration date
+        let new_targets = Targets {
+            spec_version: crate::SPEC_VERSION.to_owned(),
+            version: self.targets_version,
+            expires: self.targets_expires,
+            targets: repository.targets().signed.targets.clone(),
+            _extra: HashMap::new(),
+            // TODO - get help. Clippy wants _extra renamed to extra if we do this. Should we do so?
+            // _extra: repository.targets().signed._extra.clone(),
```
Comment (webern) [inline tuftool/src/refresh.rs]: I need some guidance here. If I 'bind' _extra by cloning the loaded _extra, then clippy wants it renamed to remove the underscore. Not sure I can avoid this by mutating the loaded object? Or should we do the rename?

Comment (webern) [reply tuftool/src/refresh.rs]: I have just seen that #20 calls _extra into question.

Comment (webern) [reply tuftool/src/refresh.rs]: Per offline convo, added `#![allow(clippy::used_underscore_binding)]` to main.rs

File: tuftool/src/root_digest.rs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - The filepath to a root.json file.
+    ///
+    ///  # Return
+    ///
+    /// * Either an Error, or a tuple consisting of `Root`, `sha256` digest, and `length`.
+    ///
+    pub(crate) fn digest_root_file(path: &PathBuf) -> Result<Self> {
```
Comment (iliana) [inline tuftool/src/root_digest.rs]: This is a little redundant, maybe rename to `fn load`?

File: tuftool/src/create.rs [hunk 12]
```diff
@@ -89,21 +91,17 @@ impl CreateArgs {
...
 
-        let root_buf = std::fs::read(&self.root).context(error::FileRead { path: &self.root })?;
-        let root = serde_json::from_slice::<Signed<Root>>(&root_buf)
-            .context(error::FileParseJson { path: &self.root })?
-            .signed;
-        let mut root_sha256 = [0; SHA256_OUTPUT_LEN];
-        root_sha256.copy_from_slice(digest(&SHA256, &root_buf).as_ref());
-        let root_length = root_buf.len() as u64;
+        let root_digest = RootDigest::digest_root_file(&self.root)?;
+        let key_pairs = keys_for_root(&self.keys, &root_digest.root)?;
```
Comment (iliana) [inline tuftool/src/create.rs]: Can `keys_for_root` become a method on `RootDigest` now?

File: tuftool/src/create.rs [hunk 13]
```diff
@@ -89,21 +91,17 @@ impl CreateArgs {
...
+        ensure!(!key_pairs.is_empty(), error::KeysNotFoundInRoot {});
```
Comment (iliana) [inline tuftool/src/create.rs]: This can maybe get rolled into `keys_for_root`.

File: tuftool/src/create.rs [hunk 14]
```diff
@@ -89,21 +91,17 @@ impl CreateArgs {
...
             args: self,
-            keys: keys_for_root(&self.keys, &root)?,
+            keys: key_pairs,
             rng: SystemRandom::new(),
-            root,
-            root_sha256,
-            root_length,
+            root: root_digest.root,
+            root_sha256: root_digest.digest,
+            root_length: root_digest.size,
```
Comment (iliana) [inline tuftool/src/create.rs]: `CreateProcess` can just take the `RootDigest` object.

File: tuftool/src/create.rs [hunk 15]
```diff
@@ -119,7 +117,7 @@ struct CreateProcess<'a> {
 }
 
 impl<'a> CreateProcess<'a> {
-    fn run(self) -> Result<()> {
+    fn run(self) -> error::Result<()> {
```
Comment (iliana) [inline tuftool/src/create.rs]: `Result` is still in scope here, no need to prepend with `error::`.

**Comment by webern** [comment]
I think the object, currently called `RootDigest` is trending towards being constructed with all the stuff that we more-or-less always need, Root, root digest, root length, random number generator, and perhaps some other bits. This object and others could then be exported in a `lib.rs` file for #43 once we are happy with the interfaces. Basically what I'm saying is that a design for #43 is starting to reveal itself.

File: tuftool/src/create.rs (hunk 13 above)
Comment (iliana) [reply tuftool/src/create.rs]: Looks like this was missed?

This error is thrown in `create` but not `sign` or `refresh` so it'd be nicet o move it into `fn load_keys`.

Comment (webern) [reply tuftool/src/create.rs]: got it this time.

**Review by zmrow — APPROVED** [review]
LGTM provided @iliana 's last concerns are fixed.

**Review by iliana — APPROVED** [review]

**Pull request merged**